flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/logical/rel/LogicalLastRow.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.logical.rel
+
+import java.util
+
+import org.apache.calcite.plan.{RelOptCluster, RelTraitSet}
+import org.apache.calcite.rel.{RelNode, SingleRel}
+
+/**
+  * Represent an upsert source.
+  *
+  * @param keyNames   The upsert key names.
+  */
+class LogicalLastRow(
```

Comment:
I still think this is a bad and misleading name for this operator. `LastRow` doesn't explain neither the purpose, nor the actual thing that it's doing. If someone was asked without prior knowledge what `LastRow` RelNode/operator does, it would be impossible to guess the correct answer. It's even more confusing when you are binding `LastRow` with "upsert" name in the java doc:

> Represent an upsert source.

> The upsert key names.

Please either change the name to something like `UpsertToRetractionsConverter` that explicitly states the purpose of this class (than you can keep `upsert` references in the java docs) or if you would like to keep naming by the thing that it does it would have to meet the following requirements:
- `LastRow` is independent of the upsert -> retraction conversion logic
- there is a reasonable chance that it will be used somewhere else outside of the upsert -> retraction conversion context 

It's a bit like we would rename `LogicalJoin` to `HashTable`. Kind of true but misleading. It could make sense if:
- `HashTable` is extracted to separate class
- `LogicalJoin` relnode/operator/concept still exists, but is just using `HashTable`
- `HashTable` is reused somewhere else

Reply:
Year, I also think `LastRow` is not perfect. However, I can't find a better one. We can't name it `UpsertToRetractionsConverter ` since we don't always convert upsert to retractions in this node. `LastRow` does explain the purpose and the thing that it is doing, i.e. get the last row from the upsert stream by the defined key.

Reply:
>  We can't name it UpsertToRetractionsConverter since we don't always convert upsert to retractions in this node. 

can you elaborate on that?

Reply:
The `LastRow` node may be a no-op node in the case such as `upsert source -> calc -> upsert sink`. And `LastRow` will convert upsert stream to retract stream if a downstream node needs it to, such as `upsert source -> calc -> retract sink`. Whether convert to retract stream will be decided by RetractionRules.